### PR TITLE
Add all the authentication queries in the plugin.

### PR DESCRIPTION
**Pull request by robinsoc**
This is a part of [story #14018](https://tuleap.net/plugins/tracker/?aid=14018) have a central management of users and groups using oauth

You can test this commit with the (future) tuleap-oauth plugin (not released yet):
 - Pull "test/tuleap-api-integration branch" from tuleap-oauth plugin. (The master branch is not compatible with this commit yet)
 - You have to install this plugin manually :
  - In this plugin, execute: mvn clean install (or mvn package)
  - Go to the tuleap-oauth plugin directory
  - mvn install:install-file -DgroupId=io.jenkins.plugins -DartifactId=tuleap-api -Dversion=1.0.3-SNAPSHOT -Dfile=path/of/your/tuleap-api.hpi_file -Dpackaging=hpi
  - mvn install:install-file -DgroupId=io.jenkins.plugins -DartifactId=tuleap-api -Dversion=1.0.3-SNAPSHOT -Dfile=path/of/your/tuleap-api.jar_file
  - mvn clean install (Test if all are ok)
 - mvn hpi:run
 - Go to the Global Configuration page to configure your Tuleap instance URL
 - Go to the Security Realm page and configure the Tuleap Authentication.
=> You should be able to be logged in Jenkins with the used user to be logged in Tuleap

Have fun :)

### CODE REVIEW

File: src/main/java/io/jenkins/plugins/tuleap_api/client/authentication/OpenIDClientApi.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import com.auth0.jwk.Jwk;
+
+import java.util.List;
+
+public interface OpenIDClientApi {
+    String DISCOVERY_API = "/.well-known/openid-configuration";
+
+    List<Jwk> getSigningKeys();
+
+    UserInfo getUserInfo(AccessToken accessToken);
```
Comment (Erwyn) [inline src/main/java/io/jenkins/plugins/tuleap_api/client/authentication/OpenIDClientApi.java:12]: Is there a reason for this not to be in the UserInfoApi ?

File: src/main/java/io/jenkins/plugins/tuleap_api/client/internals/TuleapAuthenticationApiClient.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        return new Request.Builder()
+            .url(this.tuleapConfiguration.getDomainUrl() + ACCESS_TOKEN_API)
+            .addHeader("Authorization", Credentials.basic(clientId, clientSecret.getPlainText()))
+            .addHeader("Content-Type", ACCESS_TOKEN_CONTENT_TYPE)
+            .post(requestBody)
+            .build();
+    }
+
+    @Override
+    @SuppressFBWarnings("NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE") // see https://github.com/spotbugs/spotbugs/issues/651
```
Comment (Erwyn) [inline src/main/java/io/jenkins/plugins/tuleap_api/client/internals/TuleapAuthenticationApiClient.java:100]: This one shouldn't be needed

File: src/main/java/io/jenkins/plugins/tuleap_api/client/internals/guice/JwksProvider.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import com.auth0.jwk.UrlJwkProvider;
+import com.google.inject.Inject;
+import com.google.inject.Provider;
+import io.jenkins.plugins.tuleap_server_configuration.TuleapConfiguration;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+
+
+public class JwksProvider implements Provider<UrlJwkProvider> {
```
Comment (Erwyn) [inline src/main/java/io/jenkins/plugins/tuleap_api/client/internals/guice/JwksProvider.java:12]: could we rename this to UrlJwkProviderProvider

File: src/main/java/io/jenkins/plugins/tuleap_api/client/internals/TuleapAuthenticationApiClient.java (hunk 2 above)
Comment (robinsoc) [reply src/main/java/io/jenkins/plugins/tuleap_api/client/internals/TuleapAuthenticationApiClient.java]: Ok thank you

File: src/main/java/io/jenkins/plugins/tuleap_api/client/authentication/OpenIDClientApi.java (hunk 1 above)
Comment (robinsoc) [reply src/main/java/io/jenkins/plugins/tuleap_api/client/authentication/OpenIDClientApi.java:13]: I finally decided to put all the open id query in the same interface.  And after discussion it's also ok for you

File: src/main/java/io/jenkins/plugins/tuleap_api/client/internals/TuleapAuthenticationApiAuthenticationClient.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+                throw new InvalidTuleapResponseException(response);
+            }
+            this.headerAuthenticationValidator.validateHeader(response);
+            this.accessTokenValidator.validateAccessTokenHeader(response);
+            AccessToken accessToken = this.objectMapper.readValue(Objects.requireNonNull(response.body()).string(), AccessTokenEntity.class);
+            this.accessTokenValidator.validateAccessTokenBody(accessToken);
+            return accessToken;
+        } catch (IOException | InvalidTuleapResponseException | InvalidHeaderException exception) {
+            LOGGER.severe(exception.getMessage());
+            throw new RuntimeException("Error while contacting Tuleap server", exception);
```
Comment (Erwyn) [inline src/main/java/io/jenkins/plugins/tuleap_api/client/internals/TuleapAuthenticationApiAuthenticationClient.java:84]: We could discuss this tomorrow. I think you could make them directly RuntimeExceptions

File: src/main/java/io/jenkins/plugins/tuleap_api/client/internals/entities/authentication/validators/AccessTokenValidatorImpl.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+public class AccessTokenValidatorImpl implements AccessTokenValidator {
+
+    private static final Logger LOGGER = Logger.getLogger(AccessTokenValidator.class.getName());
+
+    private static final String PRAGMA_HEADER_VALUE = "no-cache";
+
+    @Override
+    public void validateAccessTokenHeader(Response response) throws InvalidHeaderException {
+        if (!response.cacheControl().noStore()) {
+            LOGGER.severe("Bad cache policy");
```
Comment (Erwyn) [inline src/main/java/io/jenkins/plugins/tuleap_api/client/internals/entities/authentication/validators/AccessTokenValidatorImpl.java:19]: depending on what's decided about runtime exceptions, no need to log.

**Comment by Erwyn** [comment]
Almost there, could we rename it to UrlJwksProviderProvider ? I know it's two times provider but provided it provides a provider...

**Pull request merged**